### PR TITLE
src: mk: run make olddefconfig before building

### PR DESCRIPTION
**Pull request by kwy95**
Runs `make olddefconfig` before building kernel

### CODE REVIEW

File: documentation/tutorials/buildlinux.rst [hunk 1]
```diff
@@ -81,7 +81,11 @@ And it can be even simpler by just running::
 
   kw b
 
+.. note:
+    kw will also run `make olddefconfig` for you before building, so you can
```
Comment (rodrigosiqueira) [inline documentation/tutorials/buildlinux.rst:85]: Maybe:
kw runs `...` before start building...

Also, you can drop the other part that says "so you..."

File: src/mk.sh [hunk 2]
```diff
@@ -610,6 +610,9 @@ function mk_build()
     esac
   done
 
+  # sets default values for any missing configurations in .config
+  cmd_manager "$flag" "make olddefconfig"
```
Comment (rodrigosiqueira) [inline src/mk.sh:614]: Single quote around the command.

Also, `cmd_manager` will show this command; I think we can set the default flag to SILENT here. Finally, you need to check if there is a `.config` file before running this command; if there is no config file, do not run this command.

File: src/mk.sh [hunk 3]
```diff
@@ -610,6 +610,9 @@ function mk_build()
     esac
   done
 
+  # sets default values for any missing configurations in .config
```
Comment (rodrigosiqueira) [inline src/mk.sh:613]: I think we don't need this comment.

File: tests/mk_test.sh [hunk 4]
```diff
@@ -156,8 +156,9 @@ function mk_build_Test()
   assertEquals "($ID)" "$ret" "22"
 
   ID=5
-  output=$(mk_build 'TEST_MODE' | head -1) # Remove statistics output
-  expected_result="make -j$PARALLEL_CORES ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu-"
+  output=$(mk_build 'TEST_MODE' | head -2) # Remove statistics output
+  expected_result="make olddefconfig
+make -j$PARALLEL_CORES ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu-"
```
Comment (rodrigosiqueira) [inline tests/mk_test.sh:161]: I would prefer that you create an array with the command sequence. See `mk_list_remote_kernels_Test` as a reference.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by rodrigosiqueira** [comment]
@kwy95, you forgot to add tests.

File: src/mk.sh [hunk 5]
```diff
@@ -639,6 +639,10 @@ function mk_build()
     esac
   done
 
+  if [ -f "$(join_path "$PWD" '/.config')" ]; then
```
Comment (rodrigosiqueira) [inline src/mk.sh:642]: Why did you add this `/` before `.config`?

File: src/mk.sh [hunk 6]
```diff
@@ -639,6 +639,10 @@ function mk_build()
...
+    cmd_manager "${flag:-"SILENT"}" 'make olddefconfig'
```
Comment (rodrigosiqueira) [inline src/mk.sh:643]: Use single-quote around 'SILENT'

File: tests/mk_test.sh [hunk 7]
```diff
@@ -155,6 +155,22 @@ function test_mk_build()
   expected_result="make -j$PARALLEL_CORES ARCH=x86_64 "
   assertEquals "($ID)" "$expected_result" "${output}"
 
+  ID=7
+  local count=0
+  cp "$original/tests/samples/.config" .config
+  output=$(mk_build 'TEST_MODE' | head -2) # Remove statistics output
```
Comment (rodrigosiqueira) [inline tests/mk_test.sh:161]: Remove statistics output? Do we have statistics output here?

File: tests/mk_test.sh [hunk 8]
```diff
@@ -155,6 +155,22 @@ function test_mk_build()
   expected_result="make -j$PARALLEL_CORES ARCH=x86_64 "
   assertEquals "($ID)" "$expected_result" "${output}"
 
+  ID=7
```
Comment (rodrigosiqueira) [inline tests/mk_test.sh:158]: We replaced `ID` with `LINENO`.

File: tests/mk_test.sh [hunk 9]
```diff
@@ -155,6 +155,22 @@ function test_mk_build()
...
+  local count=0
+  cp "$original/tests/samples/.config" .config
+  output=$(mk_build 'TEST_MODE' | head -2) # Remove statistics output
+  expected_result=(
+    "make olddefconfig"
```
Comment (rodrigosiqueira) [inline tests/mk_test.sh:163]: single-quotes

File: tests/mk_test.sh [hunk 10]
```diff
@@ -155,6 +155,22 @@ function test_mk_build()
...
+    "make -j$PARALLEL_CORES ARCH=x86_64"
+  )
+  while read -r f; do
+    if [[ ${expected_result[$count]} != "${f}" ]]; then
+      assertEquals "($ID)" "${expected_result[$count]}" "${f}"
+    fi
+    ((count++))
+  done <<< "$output"
```
Comment (rodrigosiqueira) [inline tests/mk_test.sh:171]: We have a function for that; take a look at `compare_command_sequence`.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

File: tests/mk_test.sh (hunk 7 above)
Comment (kwy95) [reply tests/mk_test.sh:161]: Yes, `mk_build` always tries to update the statistics folder when run. Most other tests do something similar in this section, maybe we could make it so the statistics part of the code doesn't trigger while in `TEST_MODE`.

File: src/mk.sh (hunk 5 above)
Comment (kwy95) [reply src/mk.sh:642]: Oversight, thanks.

File: tests/mk_test.sh (hunk 8 above)
Comment (kwy95) [reply tests/mk_test.sh]: I left the `ID` as a comment, a lot of the tests have not been updated to use `LINENO` and I think a bigger refactor should be done to update these tests.

**Comment by rodrigosiqueira** [comment]
@kwy95 Recently, Joao break the file mk in the build.sh and deploy.sh. Could you update your PR based on the latest code in the unstable branch?

**Comment by rodrigosiqueira** [comment]
@kwy95 there is a bug in your implementation. Try `kw b`, you will see this error:
```
/home/siqueira/.local//lib/kw/kwlib.sh: line 93: SILENT: command not found
```

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Let's suppress this output:
```
#
# configuration written to .config
#
```

**Comment by rodrigosiqueira** [comment]
Also, if we don't have `.config` in the local kernel, we should not run olddefconfig.

**Comment by kwy95** [comment]
OK. I'll suppress the output, but I'm pretty sure The command checks if there is a `.config` file before executing.

**Comment by rodrigosiqueira** [comment]
Merged
